### PR TITLE
Refactor distop module 

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Refactor distop module. Currently there are two modules which can be used to distribute fp8 tensors: dist_op and fp8_op. dist_op doesn't support group and aysnc. So we need to remove the dist_op and rename fp8_op to dist_op.

**Major Revision**
- Remove old dist_op
- Rename fp8_op to dist_op
- Update document and docker file.

### CODE REVIEW

File: msamp/operators/dist_op/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+find_package(CUDA REQUIRED)
+find_package(NCCL REQUIRED)
+
+add_library(msampdist SHARED dist.cpp)
+
+target_include_directories(msampdist PUBLIC ${CUDA_INCLUDE_DIRS} ${NCCL_INCLUDE_DIRS})
+target_link_libraries(msampdist PUBLIC ${CUDA_LIBRARIES} ${CUDA_LIBRARIES})
+
+install(TARGETS msampdist)
```
Comment (wkcn) [inline msamp/operators/dist_op/CMakeLists.txt:18]: `install(TARGETS msamp_dist)` may be better.

reference: libopencv_dn.so, libsvn_ra_svn-1.so, libmpi_cxx.so

File: msamp/operators/dist_op/dist.cpp [hunk 2]
```diff
@@ -1,211 +1,55 @@
-// Reference: https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/nccl.cpp
-
-#include <ATen/ATen.h>
-#include <c10/cuda/CUDAGuard.h>
-#include <c10/util/ArrayRef.h>
-#include <c10/util/Exception.h>
-#include <c10/util/hash.h>
-#include <c10/util/irange.h>
+#include <stdio.h>
```
Comment (wkcn) [inline msamp/operators/dist_op/dist.cpp:1]: We can add our license here.
```
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
```

**Review by wkcn — COMMENTED** [review]
Thanks Yuxiang!

[resolved] After removing the old DistOp, the statement in https://github.com/Azure/MS-AMP/blob/232c40e130a3daca66ea4ca677fe6186bc45e5be/msamp/common/tensor/tensor_dist.py#L106 should be updated.

**Comment by tocean** [comment]
> Thanks Yuxiang!
> 
> After removing the old DistOp, the statement in
> 
> https://github.com/Azure/MS-AMP/blob/232c40e130a3daca66ea4ca677fe6186bc45e5be/msamp/common/tensor/tensor_dist.py#L106
> 
> should be updated.

Actually no need to update tensor_dist because the signature of DistOp::all_reduce is compatible with old DistOp::all_reduce

File: msamp/operators/dist_op/CMakeLists.txt (hunk 1 above)
Comment (tocean) [reply msamp/operators/dist_op/CMakeLists.txt]: Good suggestion. Fixed.

File: msamp/operators/dist_op/dist.cpp (hunk 2 above)
Comment (tocean) [reply msamp/operators/dist_op/dist.cpp:4]: Added.

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py [hunk 3]
```diff
@@ -462,33 +463,17 @@ def fp8_average_tensor(self, tensor):
...
-
+            # Reduce gradient.
             async_handles = []
-            grad_slice_pairs = []
             for i, (dst, bucket_offset, numel) in enumerate(rank_and_offsets):
                 grad_slice = tensor_to_reduce.narrow(0, int(bucket_offset), int(numel))
                 dst_rank = dist.get_global_rank(real_dp_process_group[i], dst)
-                fp16_grad_slice = _fp8_to_fp16(grad_slice)
-                async_handle = dist.reduce(fp16_grad_slice, dst=dst_rank, group=real_dp_process_group[i], async_op=True)
+                async_handle = DistOp.reduce(
```
Comment (wkcn) [inline msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:471]: please replace `Dtypes.kfloat8_e4m3` with `WEIGHT_GRAD_QTYPE`.
It is suitable to change the data type of weight gradient by a variable.

Comment (tocean) [reply msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:471]: Good suggestion. Fixed.

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

File: msamp/operators/dist_op/__init__.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
 
-"""Exposes the interface of MS-AMP dist_op module."""
+"""Exposes the interface of MS-AMP fp8_op module."""
```
Comment (abuccts) [inline msamp/operators/dist_op/__init__.py:4]: wrong edit?

File: tests/operators/test_dist.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        rank = self.rank
+        store = dist.FileStore(self.file_name, self.world_size)
+        dist.init_process_group(backend='nccl', store=store, rank=self.rank, world_size=self.world_size)
+        torch.cuda.set_device(rank)
+
+        tensors = [torch.rand(4).cuda(), torch.rand(4).cuda()]
+        target = tensors[0] + tensors[1]
+        tensor = tensors[rank].clone()
+
+        dist.all_reduce(tensor, op=dist.ReduceOp.SUM)
```
Comment (abuccts) [inline tests/operators/test_dist.py:52]: test other ops like avg/max as well?

**Review by abuccts — APPROVED** [review]

**Review by guoshzhao — APPROVED** [review]

File: msamp/operators/dist_op/__init__.py (hunk 4 above)
Comment (tocean) [reply msamp/operators/dist_op/__init__.py:4]: Yes, fixed.

File: tests/operators/test_dist.py (hunk 5 above)
Comment (tocean) [reply tests/operators/test_dist.py:52]: Added max and min

**Pull request merged**